runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/ConnectionIdleMonitor.kt

```diff
@@ -0,0 +1,123 @@
+import kotlin.time.measureTime
+
+@OptIn(ExperimentalOkHttpApi::class)
+internal class ConnectionIdleMonitor(val pollInterval: Duration) : ConnectionListener() {
+    private val monitors = ConcurrentHashMap<Connection, Job>()
+
+    private fun Call.callContext() =
+        request()
+            .tag(SdkRequestTag::class.java)
+            ?.callContext
+            ?: Dispatchers.IO
+
+    override fun connectionAcquired(connection: Connection, call: Call) {
+        // Non-locking map access is okay here because this code will only execute synchronously as part of a
+        // `connectionAcquired` event and will be complete before any future `connectionReleased` event could fire for
+        // the same connection.
+        monitors.remove(connection)?.let { monitor ->
+            5.seconds
+            val context = call.callContext()
+            val logger = context.logger<ConnectionIdleMonitor>()
+            logger.trace { "Cancel monitoring for $connection" }
+
+            // Use `runBlocking` because this _must_ finish before OkHttp goes to use the connection
+            val cancelTime = measureTime {
+                runBlocking(context) { monitor.cancelAndJoin() }
+            }
+
+            logger.trace { "Monitoring canceled for $connection in $cancelTime" }
+        }
+    }
+
+    override fun connectionReleased(connection: Connection, call: Call) {
+        val connId = System.identityHashCode(connection)
+        val context = call.callContext()
+        val scope = CoroutineScope(context)
+        val monitor = scope.launch(CoroutineName("okhttp-conn-monitor-for-$connId")) {
+            doMonitor(connection)
+        }
+        context.logger<ConnectionIdleMonitor>().trace { "Launched coroutine $monitor to monitor $connection" }
+
+        // Non-locking map access is okay here because this code will only execute synchronously as part of a
+        // `connectionReleased` event and will be complete before any future `connectionAcquired` event could fire for
+        // the same connection.
+        monitors[connection] = monitor
+    }
+
+    private suspend fun doMonitor(conn: Connection) {
+        val logger = coroutineContext.logger<ConnectionIdleMonitor>()
+
+        val socket = conn.socket()
+        val source = try {
+            socket.source()
+        } catch (_: SocketException) {
+            logger.trace { "Socket for $conn closed before monitoring started. Skipping polling loop." }
+            return
+        }.buffer().peek()
+
+        logger.trace { "Commence socket monitoring for $conn" }
+        var resetTimeout = true
+        val oldTimeout = socket.soTimeout
+
+        try {
+            socket.soTimeout = pollInterval.inWholeMilliseconds.toInt()
+
+            while (coroutineContext.isActive) {
+                try {
+                    logger.trace { "Polling socket for $conn" }
+                    source.readByte() // Blocking read; will take up to READ_TIMEOUT_MS to complete
```

Comment:
> will take up to READ_TIMEOUT_MS to complete

question/clarification request: I don't see a `READ_TIMEOUT_MS` / any other timeout used here in the `readByte` call